### ISSUE TITLE
NaNs popping in complicated tests

### ISSUE THREAD

**Issue by mgard** [BachiLi/redner#13]
I am trying to use Redner for a differential rendering task, and I ran into issues with more "complex" scenes. At first, I thought it was my usage that was wrong, but then I discovered I could reproduce it with the provided tests.
The simple tests (e.g. single triangle) work as expected, but the teapot tests almost always result in NaN appearing in gradients, sometimes very quickly:

```
$ python test_teapot_specular.py 
iteration: 0
loss: 24416.279296875
translation.grad: tensor([ 3.0038, -4.5837,  1.4213])
translation: tensor([19.5000,  0.5000,  1.5000], requires_grad=True)
iteration: 1
loss: 24524.2109375
translation.grad: tensor([nan, nan, nan])
translation: tensor([nan, nan, nan], requires_grad=True)
iteration: 2
```

I use the last Redner version (commit 1083be5) with Python 3.7, Pytorch 1.0, CUDA / no CUDA (tried them both), and Embree 3.4 on Anaconda on Linux.

I tried to reduce the learning rate and tweak Adam parameters, but this is still popping. Sometimes it does converge to a reasonable result, sometimes not.

I experience the same problem in my setup. I understand how tricky the hyper-parameters tuning may be and how NaNs are not a problem exclusive to redner, but would you happen to have some tricks/fixes to avoid getting those pesky NaNs, or at least make their appearance a bit more predictable?

Thanks for the great work and the efforts you make to make it available!

**Comment by HectorAnadon** [comment]
I am also getting NaNs when rendering some meshes with textures. Most of the pixel values are in the range [0-1] except a few. I will get back with a description of my setup.

**Comment by BachiLi** [comment]
It might be caused by a confusion between 0-dimensional and 1-dimensional tensor. Can you pull again from the latest commit and see if it still happens?

**Comment by mgard** [comment]
Hi @BachiLi ,
Sorry for the delay, for some reason Github did not notify me of your answer. Anyway.

I pulled the last commit and the NaNs are still popping. Tried to use `torch.autograd.detect_anomaly`, just in case, but of course since everything is in C++, the output is not useful (apart to confirm it is in the backward pass).
But now I've also discovered that on a different computer, the issue is still reproducible, but only in CPU mode (e.g. `pyredner.set_use_gpu(False)`). On CUDA, it seems to work (well, at least without NaNs in the result). I will try to investigate further.

**Comment by budmonde** [comment]
I'd suggest you put `assert(isfinite(var_name));` in the redner source code to narrow down which codepath is getting bad behaviour. From #21 and #28 I've noticed that there are potentially a lot of places in the codebase that have issues with square roots, 0/0 divisions etc. If you can narrow down what codepath is producing the original NaN that is getting propogated, we'd be able to figure out what's wrong.
